File: app/utils/LineEnding.java
/**
 * Yona, 21st Century Project Hosting SW
 * <p>
 * Copyright Yona & Yobi Authors & NAVER Corp.
 * https://yona.io
 **/
package utils;

import org.apache.commons.lang3.StringUtils;

public class LineEnding {
    public static final EndingType DEFAULT_ENDING_TYPE = EndingType.UNIX;
    public enum EndingType {
        DOS("\r\n"), UNIX("\n"), UNDEFINED("");
        public String value;

        EndingType(String lineEnding) {
            this.value = lineEnding;
        }
    }

    public static String changeLineEnding(String contents, String to){
        if(StringUtils.isNotEmpty(to) && "DOS".equalsIgnoreCase(to)){
            return changeLineEnding(contents, EndingType.DOS);
        }
        return changeLineEnding(contents, EndingType.UNIX);
    }

    public static String changeLineEnding(String contents, EndingType to){
        EndingType endingType = findLineEnding(contents);
        if (StringUtils.isEmpty(contents)) {
            return "";
        }
        if(endingType != EndingType.DOS && to == EndingType.DOS) {
            return contents.replaceAll("\n", "\n");
        }
        if(endingType != EndingType.UNIX && to == EndingType.UNIX) {
            return contents.replaceAll("\r\n", "\n");
        }
        return contents;
    }

    public static String addEOL(String contents){
        EndingType endingType = findLineEnding(contents);
        if(endingType == EndingType.UNDEFINED){
            endingType = DEFAULT_ENDING_TYPE;
        }
        if ( contents != null ){
            if(!contents.endsWith(endingType.value)){
                return contents.concat(endingType.value);
            }
        }
        return contents;
    }

    public static EndingType findLineEnding(String contents){
        if (StringUtils.isEmpty(contents)) {
            return EndingType.UNDEFINED;
        }
        
        if(contents.contains("\r\n") ) {
            return EndingType.DOS;
        } else {
            return EndingType.UNIX;
        }
    }
}

File: src/main/kotlin/com/github/yonaprojects/yona/domain/support/LineEnding.kt
package com.github.yonaprojects.yona.domain.support

// yona utils/LineEnding.java 대응(코드브라우저 편집 커밋에서 post.lineEnding 처리에 필요).
// changeLineEnding(contents, EndingType.DOS) 분기의 `contents.replace("\n", "\n")`는 yona 원본의
// `contents.replaceAll("\\n", "\\n")`과 동일한 무의미한 치환(no-op) 버그를 그대로 재현한 것 —
// DOS 개행 변환이 실제로는 동작하지 않는 legacy 동작을 의도적으로 보존한다.
object LineEnding {
    enum class EndingType(val value: String) {
        DOS("\r\n"), UNIX("\n"), UNDEFINED("")
    }

    val DEFAULT_ENDING_TYPE = EndingType.UNIX

    fun changeLineEnding(contents: String, to: String?): String {
        return if (!to.isNullOrEmpty() && to.equals("DOS", ignoreCase = true)) {
            changeLineEnding(contents, EndingType.DOS)
        } else {
            changeLineEnding(contents, EndingType.UNIX)
        }
    }

    fun changeLineEnding(contents: String, to: EndingType): String {
        val endingType = findLineEnding(contents)
        if (contents.isEmpty()) {
            return ""
        }
        if (endingType != EndingType.DOS && to == EndingType.DOS) {
            return contents.replace("\n", "\n")
        }
        if (endingType != EndingType.UNIX && to == EndingType.UNIX) {
            return contents.replace("\r\n", "\n")
        }
        return contents
    }

    fun addEOL(contents: String?): String? {
        if (contents == null) {
            return contents
        }
        var endingType = findLineEnding(contents)
        if (endingType == EndingType.UNDEFINED) {
            endingType = DEFAULT_ENDING_TYPE
        }
        if (!contents.endsWith(endingType.value)) {
            return contents + endingType.value
        }
        return contents
    }

    fun findLineEnding(contents: String?): EndingType {
        if (contents.isNullOrEmpty()) {
            return EndingType.UNDEFINED
        }
        return if (contents.contains("\r\n")) EndingType.DOS else EndingType.UNIX
    }
}
